### PR TITLE
Us31 confirm reservation

### PR DESCRIPTION
**Pull request by JoJ3o**

### CODE REVIEW

File: api/src/main/java/com/buurbak/api/reservations/service/ReservationService.java [hunk 1]
```diff
@@ -24,37 +30,87 @@ public class ReservationService {
...
     private final CustomerService customerService;
     private final TrailerOfferService trailerOfferService;
+    private final ReservationEmailService reservationEmailService;
 
     public Reservation getReservation(UUID id) throws ReservationNotFoundException {
         return reservationRepository.findById(id).orElseThrow(ReservationNotFoundException::new);
     }
 
-    public Reservation addReservation(ReservationDTO reservationDTO, String username) throws CustomerNotFoundException, TrailerOfferNotFoundException {
+    public Reservation addReservation(ReservationDTO reservationDTO, String username, HttpServletRequest request) throws CustomerNotFoundException, TrailerOfferNotFoundException, ReservationAlreadyProgressedException, ReservationRenterIsOwnerException, MessagingException {
```
Comment (Alyvox) [inline api/src/main/java/com/buurbak/api/reservations/service/ReservationService.java:39]: Add check to see if there is already another reservation on that time

File: api/src/test/java/com/buurbak/api/reservations/ReservationIntegrationTest.java [hunk 2]
```diff
@@ -43,7 +43,8 @@ public class ReservationIntegrationTest {
     void setUp() {
         mapper = new ObjectMapper();
         mapper.registerModule(new JavaTimeModule());
-        reservationId = reservationRepository.save(rdg.reservation.nextObject(Reservation.class)).getId();
+        Reservation randomReservation = rdg.reservation.nextObject(Reservation.class);
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/ReservationIntegrationTest.java:46]: You could add .getId() behind the generation

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 3]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+    @Test
+    void confirmReservation_shouldSetReservationToConfirmedAndCallSaveAndSendConfirmedMails() throws MessagingException {
+        Customer renter = new Customer();
+        Customer owner = new Customer();
+        UUID trailerId = UUID.randomUUID();
+        LocalDateTime startTime = LocalDateTime.now();
+        LocalDateTime endTime = LocalDateTime.now();
+        LocalDateTime lastChanged = LocalDateTime.now();
+
+        Reservation reservation = new Reservation();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:138]: Use easyRandom to stay consistent with generating data

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 4]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+
+        Assertions.assertTrue(reservation.getConfirmed());
+        verify(reservationRepository).save(reservation);
+        verify(reservationEmailService).sendConfirmedMails(reservation.getId(), request, renter, owner, trailerId, startTime, endTime, lastChanged);
+    }
+
+    @Test
+    void confirmReservation_ShouldThrowReservationAlreadyProgressedException_WhenLastChangedDoesNotMatchReservationsUpdatedAt() throws ReservationAlreadyProgressedException, MessagingException {
+        UUID id = UUID.randomUUID();
+        Reservation reservation = new Reservation();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:162]: Use easyRandom to stay consitant with generating data

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 5]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+        assertThrows(ReservationAlreadyProgressedException.class, () -> service.confirmReservation(id, lastChanged, request));
     }
-}
\ No newline at end of file
+
+    @Test
+    void confirmReservation_ShouldThrowReservationAlreadyProgressedException_WhenReservationIsAlreadyCancelled() throws ReservationAlreadyProgressedException, MessagingException {
+        UUID id = UUID.randomUUID();
+        LocalDateTime lastChanged = LocalDateTime.now();
+        Reservation reservation = new Reservation();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:175]: Use easyRandom to stay consitant with generating data

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 6]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+        reservation.setConfirmed(false);
+
+        when(reservationRepository.findById(id)).thenReturn(Optional.of(reservation));
+
+        assertThrows(ReservationAlreadyProgressedException.class, () -> service.confirmReservation(id, lastChanged, request));
+    }
+
+    @Test
+    void cancelReservation_shouldSendRenterCancelMails_whenCancelledByRenter() throws MessagingException {
+        UUID id = UUID.randomUUID();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:186]: Use easyRandom to stay consitant with generating data

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 7]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+
+        assertFalse(reservation.getConfirmed());
+        assertEquals("renter", reservation.getCanceledBy());
+        verify(reservationRepository).save(reservation);
+        verify(reservationEmailService).sendRenterCancelMails(renterEmail, ownerEmail);
+    }
+
+    @Test
+    void cancelReservation_shouldSendOwnerCancelMails_whenCancelledByOwner() throws MessagingException {
+        UUID id = UUID.randomUUID();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:213]: Use easyRandom to stay consitant with generating data

File: api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java [hunk 8]
```diff
@@ -80,28 +102,163 @@ void shouldUpdateReservation() {
...
+
+        assertFalse(reservation.getConfirmed());
+        assertEquals("owner", reservation.getCanceledBy());
+        verify(reservationRepository).save(reservation);
+        verify(reservationEmailService).sendOwnerCancelMails(renterEmail, ownerEmail);
+    }
+
+    @Test
+    void cancelReservation_shouldSendDeniedMails_whenReservationIsNotConfirmed() throws MessagingException {
+        UUID id = UUID.randomUUID();
```
Comment (Alyvox) [inline api/src/test/java/com/buurbak/api/reservations/service/ReservationServiceTest.java:240]: Use easyRandom to stay consitant with generating data

**Review by Alyvox — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/buurbak/api/reservations/service/ReservationService.java (hunk 1 above)
Comment (JoJ3o) [reply api/src/main/java/com/buurbak/api/reservations/service/ReservationService.java:39]: Added to backlog

File: api/src/test/java/com/buurbak/api/reservations/ReservationIntegrationTest.java (hunk 2 above)
Comment (JoJ3o) [reply api/src/test/java/com/buurbak/api/reservations/ReservationIntegrationTest.java]: Fixed in latest commit

**Review by Alyvox — APPROVED** [review]

**Pull request merged**